### PR TITLE
Create a documentation (in /docs) about testing TEAL

### PR DESCRIPTION
**Pull request by ratik21**
Added documentation describing
- What is TEAL
- How TEAL works
- Example execution - Stateless and Stateful TEAL
- What we support and what we don't
- Links to examples

### CODE REVIEW

File: docs/testing-teal.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Testing TEAL
+
+[TEAL](https://developer.algorand.org/docs/reference/teal/specification/) is a bytecode based stack language that executes inside Algorand transactions to check the parameters of the transaction and approve the transaction as if by a signature. `@algorand-builder/algorand-js` provides a simple workflow to execute TEAL code with high speed and throughput (as the transactions are not processed on the blockchain but in a test environment).
```
Comment (robert-zaremba) [inline docs/testing-teal.md:3]:
```suggestion
[TEAL](https://developer.algorand.org/docs/reference/teal/specification/) is a bytecode based stack language that executes inside Algorand transactions to check the parameters of the transaction and approve the transaction as if by a signature. `@algorand-builder/algorand-js` provides a lightweight runtime and TEAL interpreter to test Algorand Smart Contracts.
```

File: docs/testing-teal.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+**NOTE:** TEAL cannot modify or create transactions, only reject or approve them. Approval is signaled by finishing with the stack containing a single non-zero uint64 value.
```
Comment (robert-zaremba) [inline docs/testing-teal.md:5]:
```suggestion
**NOTE:** TEAL can not modify nor create transactions, only reject or approve them. Approval is signaled by finishing with the stack containing a single non-zero uint64 value.
```

File: docs/testing-teal.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+## How it works
+
+`@algorand-builder/algorand-js` (JavaScript Algorand runtime) executes transactions and processes TEAL in 3 parts :-
+
+- [Runtime](../packages/algorand-js/src/runtime): For a transaction or txn group, the state management is handled by the `Runtime`. User can use `Runtime` class to set up accounts, create applications, opt-in to app, update app..etc
```
Comment (robert-zaremba) [inline docs/testing-teal.md:11]:
```suggestion
- [Runtime](../packages/algorand-js/src/runtime): For a transaction or txn group, the state management is handled by the `Runtime`. User can use `Runtime` object to set up accounts, create applications, opt-in to app, update app, etc...
```

File: docs/testing-teal.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+- [Parser](../packages/algorand-js/src/parser): Reads TEAL code and converts it to a list of opcodes which are executable by the interpreter. If any opcode/data in teal code is invalid, parser will throw an error.
+- [Interpreter](../packages/algorand-js/src/interpreter): Executes the list of opcodes returned by parser and updates stack after each execution. After the end of execution, if top of stack contains a non-zero uint64 value then the teal code is approved, and transaction can be executed.
```
Comment (robert-zaremba) [inline docs/testing-teal.md:13]:
```suggestion
- [Interpreter](../packages/algorand-js/src/interpreter): Executes the list of opcodes returned by parser and updates stack after each execution. At the end of execution, if the stack contains a single non-zero uint64 element then the teal code is approved, and transaction can be executed.
```

File: docs/testing-teal.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+## Run tests
+In this section we will demonstrating executing transactions with stateless and stateful teal.
```
Comment (robert-zaremba) [inline docs/testing-teal.md:16]:
```suggestion
In this section we will demonstrate executing transactions with stateless and stateful teal.
```

File: docs/testing-teal.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+### Stateless TEAL
+
+Let's try to execute a transaction where the user (say `john`) can withdraw funds from an `escrow` account based on a stateless smart contract logic. Teal code can be found [here](../packages/algorand-js/test/fixtures/escrow-account/assets/escrow.teal).
```
Comment (robert-zaremba) [inline docs/testing-teal.md:20]:
```suggestion
Let's try to execute a transaction where a user (say `john`) can withdraw funds from an `escrow` account based on a stateless smart contract logic. TEAL code can be found [here](../packages/algorand-js/test/fixtures/escrow-account/assets/escrow.teal).
```

File: docs/testing-teal.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+**NOTE:** TEAL cannot modify or create transactions, only reject or approve them. Approval is signaled by finishing with the stack containing a single non-zero uint64 value.
+
+## How it works
+
+`@algorand-builder/algorand-js` (JavaScript Algorand runtime) executes transactions and processes TEAL in 3 parts :-
+
+- [Runtime](../packages/algorand-js/src/runtime): For a transaction or txn group, the state management is handled by the `Runtime`. User can use `Runtime` class to set up accounts, create applications, opt-in to app, update app..etc
+- [Parser](../packages/algorand-js/src/parser): Reads TEAL code and converts it to a list of opcodes which are executable by the interpreter. If any opcode/data in teal code is invalid, parser will throw an error.
+- [Interpreter](../packages/algorand-js/src/interpreter): Executes the list of opcodes returned by parser and updates stack after each execution. After the end of execution, if top of stack contains a non-zero uint64 value then the teal code is approved, and transaction can be executed.
+
```
Comment (robert-zaremba) [inline docs/testing-teal.md:14]: Please add also a documentation about `StoreAccountImpl`.

Comment (robert-zaremba) [reply docs/testing-teal.md:14]: Note, we should rename StoreAccountImpl: https://www.pivotaltracker.com/story/show/176449616

File: docs/testing-teal.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+- First let's set up the state: initialize accounts and set up runtime (snippet from mocha test is also provided below).
+```
+const escrow = new StoreAccountImpl(1000e6); // 1000 ALGO
+const john = new StoreAccountImpl(500, johnAccount); // 0.005 ALGO
+const runtime = new Runtime([escrow, john]); // setup runtime
+```
+![image](https://user-images.githubusercontent.com/33264364/104197847-81afe200-544b-11eb-8553-b2446ea9c763.png)
+
+- Execute transaction (using `runtime.executeTx()`) with valid txnParams.
+![image](https://user-images.githubusercontent.com/33264364/104199023-f33c6000-544c-11eb-975b-1f8f72508ee6.png)
```
Comment (robert-zaremba) [inline docs/testing-teal.md:30]: Let's use code sections. It's easier to edit it in the future.

File: docs/testing-teal.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+In test, we are first checking the initial balance which we set during initialization. Then we are executing transaction by passing `txnParams, teal code (as string), external arguments to smart contract`. After execution, we are verifying the account balances if the funds are withrawn from `escrow`.
```
Comment (robert-zaremba) [inline docs/testing-teal.md:31]:
```suggestion
In test, we are first checking the initial balance which we set during initialization. Then we are executing transaction by passing `txnParams, teal code (as a string), external arguments to smart contract`. After execution, we are verifying the account balances if the funds are withdrawn from `escrow`.
```

File: docs/testing-teal.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+- Executing transaction with invalid txnParams results in failure.
+![image](https://user-images.githubusercontent.com/33264364/104201136-6cd54d80-544f-11eb-80ea-9ca307bc189d.png)
+
+Full mocha test with more transactions can be found [here](../packages/algorand-js/test/integration/escrow-account.ts).
+
+### Stateful TEAL
+
+Now, we will execute a transaction with stateful TEAL (which increments a global and local "counter" by on each application call). Teal code can be found [here](../packages/algorand-js/test/fixtures/stateful/assets/counter-approval.teal)
+
+- Similar to above test, we need to setup accounts and initialize runtime. Now, for stateful txn, we also need to create a new application in user account and also do opt-in (to call the stateful smart contract later). User can use `runtime.addApp()` and `runtime.optInToApp()` for app setup.
```
Comment (robert-zaremba) [inline docs/testing-teal.md:42]:
```suggestion
- Similar to the previous test, we need to setup accounts and initialize runtime. Now, for stateful smart contract, we also need to create a new application in user account and opt-in (to call the stateful smart contract later). User can use `runtime.addApp()` and `runtime.optInToApp()` for app setup.
```

File: docs/testing-teal.md [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+![image](https://user-images.githubusercontent.com/33264364/104204711-94c6b000-5453-11eb-99e5-f772ce4b5c92.png)
```
Comment (robert-zaremba) [inline docs/testing-teal.md:43]: ditto, let's use code snippet, not an image.

File: docs/testing-teal.md [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+
+- After set up, let's call the stateful smart contract and check the updated global state
+![image](https://user-images.githubusercontent.com/33264364/104205377-5382d000-5454-11eb-8b57-8a6690694c61.png)
+In this test, after executing the transaction (call stateful smart contract), we are verifying if the `global state` and `local state` is updated. User can use `runtime.getGlobalState()` and `account.getLocalState()` to check state.
```
Comment (robert-zaremba) [inline docs/testing-teal.md:47]:
```suggestion
In this test, after executing the transaction (stateful smart contract call), we are verifying if the `global state` and `local state` is updated. User can use `runtime.getGlobalState()` and `account.getLocalState()` to check state.
```

**Review by robert-zaremba — APPROVED** [review]
pre approving. Left few comments and suggestions. Please use code snippets instead of images.

File: docs/testing-teal.md (hunk 7 above)
Comment (ratik21) [reply docs/testing-teal.md:14]: ok

File: docs/testing-teal.md (hunk 11 above)
Comment (ratik21) [reply docs/testing-teal.md:43]: understood, thanks for note

File: docs/testing-teal.md (hunk 8 above)
Comment (ratik21) [reply docs/testing-teal.md:30]: ok

**Pull request merged**